src/relative.rs

```diff
@@ -0,0 +1,204 @@
+
+// Taken from the pathdiff crate, which adapted the original rustc's path_relative_from
+// https://github.com/Manishearth/pathdiff/blob/master/src/lib.rs
+// https://github.com/rust-lang/rust/blob/e1d0de82cc40b666b88d4a6d2c9dcbc81d7ed27f/src/librustc_back/rpath.rs#L116-L158
+
+use std::path::*;
+
+/// Provides helper methods on Path and PathBuf to for creating RelativePath.
+pub trait PathExt {
+    /// Make a relative path from a provided root directory path.
+    ///
+    /// This may not function correctly if either path includes a symlink. Resolve or discard
+    /// symlink paths before making them relative.
+    ///
+    /// ```rust
+    /// use relative_path::PathExt;
+    /// use std::path::*;
+    ///
+    /// let baz = PathBuf::from("/foo/bar/baz");
+    /// let bar = PathBuf::from("/foo/bar");
+    /// let quux = PathBuf::from("/foo/bar/quux");
+    /// assert_eq!(bar.relative_to(&baz), Ok("../".into()));
+    /// assert_eq!(baz.relative_to(&bar), Ok("baz".into()));
+    /// assert_eq!(quux.relative_to(&baz), Ok("../quux".into()));
+    /// assert_eq!(baz.relative_to(&quux), Ok("../baz".into()));
+    /// assert_eq!(bar.relative_to(&quux), Ok("../".into()));
+    ///
+    /// assert_eq!(baz.as_path().relative_to(&bar), Ok("baz".into()));
+    /// assert_eq!(baz.as_path().relative_to(bar.as_path()), Ok("baz".into()));
+    /// ```
+    fn relative_to(&self, root: impl AsRef<Path>) -> Result<RelativePathBuf, FromPathError>;
+}
+
+impl PathExt for Path {
+    fn relative_to(&self, root: impl AsRef<Path>) -> Result<RelativePathBuf, FromPathError> {
+        let root = root.as_ref();
+        if self.is_absolute() != root.is_absolute() {
+            Err(FromPathErrorKind::NonRelative.into())
+        } else {
+            let mut path_comps = self.components();
+            let mut root_comps = root.components();
+            let mut comps: Vec<Component> = vec![];
+            loop {
+                match (path_comps.next(), root_comps.next()) {
+                    (None, None) => break,
+                    (Some(a), None) => {
+                        comps.push(a);
+                        comps.extend(path_comps.by_ref());
+                        break;
+                    }
+                    (None, _) => comps.push(Component::ParentDir),
+                    (Some(a), Some(b)) if comps.is_empty() && a == b => (),
+                    (Some(a), Some(b)) if b == Component::CurDir => comps.push(a),
+                    (Some(_), Some(b)) if b == Component::ParentDir => {
+                        return Err(FromPathErrorKind::NonRelative.into())
+                    } // TODO - Correct error?
+                    (Some(a), Some(_)) => {
+                        comps.push(Component::ParentDir);
+                        for comp in root_comps {
+                            match comp {
+                                Component::ParentDir => {
+                                    comps.pop().expect("Root was relative?");
```

Comment:
Raise an error here instead? The panic implies that this code is unreachable, and if so I'd like to see that more thoroughly documented. Else just raise an error so we don't accidentally introduce panicking paths now or in the future.

Reply:
Okay, I was able to trigger the panic actually with something like: `PathBuf::from(".").relative_to(PathBuf::from("a/../.."))`. I've gone ahead and updated it to return NonRelative. It isn't the best error type though for this, and a few other things; it may be worth it to create a separate error enum for the creation of relative paths